Avoid unpinning in-use prog arrays in init.sh

ti-mo opened the pull request:
This was tested by adding `printk("missed tail call %d\n", index);` before the end of `ep_tail_call()`. I also added a `sleep 10` between the calls to `bpf_load` on `bpf_overlay.c` in `init.sh` to exacerbate the issue, otherwise the failure was only reproducible in VirtualBox. Cilium was also running with `resources.limits.cpu=400m` to slow down execution.

During my tests, the only dropped calls I observed this way were to `CILIUM_CALL_DROP_NOTIFY` while replacing `bpf_overlay.o`. The changes in this PR are the only remaining cases that stood out to me, and observing `tc/globals` and `bpftool map list` during agent upgrade seems to display correct behaviour at least for the features enabled in minikube.

This will need to be backported to fix `K8sUpdates`, as downgrading to a stable 1.11 version of Cilium will still unpin prog arrays too early, failing the test.

Fixes: #16561

Addresses #17758 but needs backport to hopefully fix it.

```release-note
Avoid deleting in-use program arrays in bpf_load() and bpf_load_cgroups() in init.sh
```

cc @brb @pchaigno

pchaigno commented on bpf/init.sh:
```diff
@@ -445,8 +447,16 @@ if [ "${TUNNEL_MODE}" != "<nil>" ]; then
+
+	if ! bpf_load "$ENCAP_DEV" "$COPTS" egress bpf_overlay.c bpf_overlay.o to-overlay "${CALLS_MAP}"; then
+		cilium bpf migrate-maps -e bpf_overlay.o -r 1
+	fi
```
I think we are missing error cases here. If the load fails on ingress or egress, we want to exit the script right after running `cilium bpf migrate-maps`.

There's the same issue with all similar `if ! bpf_load` changes.

ti-mo replied:
Ah yes, thanks!

pchaigno commented on bpf/init.sh:
```diff
@@ -445,8 +447,16 @@ if [ "${TUNNEL_MODE}" != "<nil>" ]; then
-	bpf_load $ENCAP_DEV "$COPTS" "egress" bpf_overlay.c bpf_overlay.o to-overlay ${CALLS_MAP}
+
+	if ! bpf_load "$ENCAP_DEV" "$COPTS" ingress bpf_overlay.c bpf_overlay.o from-overlay "${CALLS_MAP}"; then
+		cilium bpf migrate-maps -e bpf_overlay.o -r 1 && exit 1
```
Sorry, should have noticed this earlier, but couldn't we fold this into `bpf_load` (same for `bpf_load_cgroup`)? It would make most changes unnecessary and reduce code duplication.
```diff
- cilium bpf migrate-maps -e $OUT -r $RETCODE
+ if [ "$RETCODE" -ne 0 ]; then
+   cilium bpf migrate-maps -e bpf_overlay.o -r $RETCODE
+ fi
```

ti-mo replied:
Ah, yes. :) No need to even capture the return code if migration reverts are always done within those functions. Also removed the use of `set`. Thanks!

ti-mo committed:
bpf/init.sh: lift map migration out of bpf_load() and bpf_load_cgroups()

To avoid missed tail calls during agent upgrade when prog arrays are resized,
make sure all bpf entrypoints belonging to an interface are attached before
unpinning the old prog arrays.

Changed the control flow of bpf_load(_cgroups) as to not require `set`.

Signed-off-by: Timo Beckers <timo@isovalent.com>

brb reviewed (approved):
Thanks!
